lib/puppetserver/ca/list_action.rb

```diff
@@ -0,0 +1,155 @@
+          opts.on('--all', 'List all certificates') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def run(input)
+        config = input['config']
+
+        if config
+          errors = FileUtilities.validate_file_paths(config)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        puppet = PuppetConfig.parse(config)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        all_certs = get_all_certs(puppet.settings)
+        return 1 if all_certs.nil?
+
+        requested, signed, revoked = separate_certs(all_certs)
+        input['all'] ? output_certs_by_state(requested, signed, revoked) : output_certs_by_state(requested)
+
+        return 0
+      end
+
+      def output_certs_by_state(requested, signed = [], revoked = [])
+
+        if revoked.empty? && signed.empty? && requested.empty?
+          @logger.inform "No certificates to list"
+          return
+        end
+
+        unless requested.empty?
+          @logger.inform "Requested Certificates:"
+          output_certs(requested)
+        end
+
+        unless signed.empty?
+          @logger.inform "Signed Certificates:"
+          output_certs(signed)
+        end
+
+        unless revoked.empty?
+          @logger.inform "Revoked Certificates:"
+          output_certs(revoked)
+        end
+      end
+
+      def output_certs(certs)
+        padded = 0
+        certs.each do |cert|
+          cert_size = cert["name"].size
+          padded = cert_size if cert_size > padded
+        end
+
+        certs.each do |cert|
+          @logger.inform "    #{cert["name"]}".ljust(padded + 6) + " (SHA256) " + " #{cert["fingerprints"]["SHA256"]}" +
+                             (cert["dns_alt_names"].empty? ? "" : "\talt names: #{cert["dns_alt_names"]}")
+          end
+      end
+
+      def http_client(settings)
+        @client ||= HttpClient.new(settings[:localcacert],
+                                   settings[:certificate_revocation],
+                                   settings[:hostcrl])
+      end
+
+      def get_certificate_statuses(settings)
+        client = http_client(settings)
+        url = client.make_ca_url(settings[:ca_server],
+                                 settings[:ca_port],
+                                 'certificate_statuses',
+                                 'any_key')
+        client.with_connection(url) do |connection|
+          connection.get(url)
+        end
+      end
+
+      def separate_certs(all_certs)
+        certs = all_certs.group_by { |v| v["state"]}
+        requested = certs.fetch("requested", [])
+        signed = certs.fetch("signed", [])
+        revoked = certs.fetch("revoked", [])
+        return requested, signed, revoked
+      end
+
+      def get_all_certs(settings)
+        result = get_certificate_statuses(settings)
+
+        unless result.code == '200'
+          @logger.err 'Error:'
+          @logger.err "    #{result.inspect}"
```

Comment:
maybe like we've done elsewhere?  I'll leave it up to you though
```
      @logger.err "      code: #{result.code}"
      @logger.err "      body: #{result.body.to_s}" if result.body
```

Reply:
yep, will do